core/src/main/java/org/apache/iceberg/ReplaceManifests.java

```diff
@@ -0,0 +1,217 @@
+    return this;
+  }
+
+  @Override
+  protected Map<String, String> summary() {
+    return summaryBuilder.build();
+  }
+
+  @Override
+  public ReplaceManifests clusterBy(Function<DataFile, Object> func) {
+    this.clusterByFunc = func;
+    return this;
+  }
+
+  @Override
+  public ReplaceManifests filter(Function<ManifestFile, Boolean> func) {
+    this.filterFunc = func;
+    return this;
+  }
+
+  @Override
+  public List<ManifestFile> apply(TableMetadata base) {
+    Preconditions.checkNotNull(clusterByFunc, "clusterBy function cannot be null");
+
+    List<ManifestFile> existingManifests = Collections.synchronizedList(new ArrayList<>());
+
+    try {
+      Tasks.foreach(base.currentSnapshot().manifests())
+          .executeWith(ThreadPools.getWorkerPool())
+          .run(manifest -> {
+            if (filterFunc != null && !filterFunc.apply(manifest)) {
+              existingManifests.add(manifest);
+            } else {
+              long entryNum = manifest.addedFilesCount() + manifest.existingFilesCount() + manifest.deletedFilesCount();
+              long avgEntryLen = manifest.length() / entryNum;
+
+              try (ManifestReader reader =
+                     ManifestReader.read(ops.io().newInputFile(manifest.path()))) {
+                FilteredManifest filteredManifest = reader.select(Lists.newArrayList("*"));
+                filteredManifest.iterator().forEachRemaining(
+                    file -> appendFile(file, avgEntryLen, clusterByFunc.apply(file))
+                );
+
+              } catch (IOException x) {
+                throw new RuntimeIOException(x);
+              }
+            }
+          });
+    } finally {
+      Tasks.foreach(writers.values()).executeWith(ThreadPools.getWorkerPool()).run(writer -> writer.close());
+    }
+
+    newManifests.addAll(
+        writers.values()
+          .stream()
+          .map(WriterWrapper::toManifestFile)
+          .collect(Collectors.toList())
+    );
+
+    // put new manifests at the beginning
+    List<ManifestFile> apply = new ArrayList<>();
+    apply.addAll(newManifests);
+    apply.addAll(existingManifests);
+
+    return apply;
+  }
+
+  private void appendFile(DataFile file, long avgEntryLen, Object key) {
+    Preconditions.checkNotNull(file, "Data file cannot be null");
+    Preconditions.checkNotNull(key, "Key cannot be null");
+
+    WriterWrapper writer = getWriter(key);
+    writer.addFile(file, avgEntryLen);
+  }
+
+  private WriterWrapper getWriter(Object key) {
+    WriterWrapper writer = writers.get(key);
+    if (writer == null) {
+      synchronized (writers) {
+        writer = writers.get(key); // check again after getting lock
+        if (writer == null) {
+          writer = new WriterWrapper();
+          writers.put(key, writer);
+        }
+      }
+    }
+    return writer;
+  }
+
+  @Override
+  protected void cleanAll() {
+    writers.values().stream().forEach(writer -> writer.close());
```

Comment:
I don't think you need to implement cleanAll to close the writers. This won't be called while apply is running and apply closes all of its writers.

Reply:
True. I removed this